## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [46]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [47]:
# View the column names in the data
column_names = df.columns
print(column_names)



Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [48]:
# Use the describe function to gather some basic statistics
summary_stats = df.describe()
summary_stats



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [49]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.info()
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [50]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3).rename('category')
top_categories.index.name = None
top_categories


consumables    23538
furniture      11915
software        8400
Name: category, dtype: int64

In [51]:
# For the category with the most entries,
# which subcategory had the most entries?
top_category = df['category'].value_counts().idxmax()
top_subcategory_series = df[df['category'] == top_category]['subcategory'].value_counts()

top_subcategory = top_subcategory_series.idxmax()
top_subcategory_count = top_subcategory_series.max()

print(f"{top_subcategory} {top_subcategory_count}\nname: subcategory, dtype:int64")


bathroom supplies 6424
name: subcategory, dtype:int64


In [52]:
# Which five clients had the most entries in the data?
top_clients = df['client_id'].value_counts().head(5)
top_clients.index.name = None
top_clients


33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [53]:
# Store the client ids of those top 5 clients in a list.
top_clients = df['client_id'].value_counts().head(5)
top_client_ids = top_clients.index.tolist()
top_client_ids


[33615, 66037, 46820, 38378, 24741]

In [54]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client_id = df['client_id'].value_counts().idxmax()
top_client_entries = df[df['client_id'] == top_client_id]

total_units_ordered = top_client_entries['qty'].sum()
total_units_ordered


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [55]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['line_subtotal'] = df['unit_price']  * df['qty']
columns_to_display = ['unit_price','qty', 'line_subtotal']
df[columns_to_display].head()


unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [56]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = df.apply(lambda row: 7 * row['total_weight'] if row ['total_weight'] > 50 else 10 * row['total_weight'], axis=1)
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head()

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [57]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df.apply(lambda row: 7 * row['total_weight'] if row['total_weight'] > 50 else 10 * row['total_weight'], axis=1)
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head()



unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [71]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['line_price'] = (df['line_subtotal'] + df['shipping_price'] + (df['line_subtotal'] + df['shipping_price']) * 0.0925).round(2)
df[['line_subtotal', 'shipping_price', 'line_price']].head()
df['line_profit'] = df['line_price'] - df['line_cost']


In [73]:
# Create a column for the profit of
# each line using line cost and line price
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
desired_columns = [
        'first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id', 'order_date', 'order_week', 'order_year', 'unit_price', 'unit_cost', 'unit_weight', 'qty', 'line_number', 'line_subtotal', 'total_weight', 'shipping_price', 'line_price', 'line_cost', 'line_profit'
]
df[desired_columns].head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...  unit_cost  unit_weight  qty  line_number  \
0          17        2023  ...     762.71         7.50  105            1   
1          20        2023  ...      15.09         1.49   21            0   
2           4        2023  ...       7.86         1.68   39            6   
3          17        2023  ...      24.85         1.23   29            3   
4          21        2023  ...     108.17        46.43   20            1   

   line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0      115164.00        787.50         5512.50   131839.08   85597.05   
1         523.95         31.29          312.90      914.26     629.79   
2         527.28         65.52          458.64     1077.12     765.18   
3        1056.18         35.67          356.70     1543.57    1077.35   
4        3902.00        928.60         6500.20    11364.40    8663.60   

   line_profit  
0     46242.03  
1       284.47  
2       311.94  
3       466.22  
4      2700.80  

[5 rows x 21 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [74]:
# Check your work using the totals above
order_id_1 = 2742071
order_id_2 = 2173913
order_id_3 = 6128929

total_price_1 = df[df['order_id'] == order_id_1]['line_price'].sum()
total_price_2 = df[df['order_id'] == order_id_2]['line_price'].sum()
total_pride_3 = df[df['order_id'] == order_id_3]['line_price'].sum()

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [75]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
top_clients = df['client_id'].value_counts().head(5).index.tolist()
top_clients_df = df[df['client_id'].isin(top_clients)]
total_spending_by_client = top_clients_df.groupby('client_id')['line_price'].sum()
total_spending_by_client.index.name = None
total_spending_by_client.apply(lambda x: f"${x:.2f}")


24741    $82268892.02
33615     $8377308.52
38378    $12906550.87
46820     $9743794.36
66037    $10259514.79
Name: line_price, dtype: object

In [77]:
top_clients_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number', 'line_subtotal', 'total_weight', 'shipping_price',
       'line_price', 'line_cost', 'line_profit'],
      dtype='object')

In [78]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

top_clients = df['client_id'].value_counts().head(5).index.tolist()
top_clients_df = df[df['client_id'].isin(top_clients)]

client_summary = top_clients_df.groupby('client_id').agg(
    qty=('qty', 'sum'),
    shipping_price=('shipping_price', 'sum'),
    line_price=('line_price', 'sum'),
    line_cost=('line_cost', 'sum'),
    line_profit=('line_profit', 'sum')
)
client_summary = client_summary.sort_values(by='line_profit', ascending=False)
client_summary


qty  shipping_price   line_price    line_cost  line_profit
client_id                                                               
24741      239862      5126448.37  82268892.02  45688899.71  36579992.31
38378       73667      3429455.40  12906550.87   9634720.98   3271829.89
66037       43018      1395151.85  10259514.79   7004482.98   3255031.81
46820       75768      1601448.84   9743794.36   7007191.64   2736602.72
33615       64313      1828984.89   8377308.52   6175313.91   2201994.61

In [79]:
top_clients_df

first     last                                           job  \
5        Jessica    Reyes  Production designer, theatre/television/film   
12       Jessica    Reyes  Production designer, theatre/television/film   
137        Bryan    Myers                                Chief of Staff   
214        Bryan    Myers                                Chief of Staff   
216       Angela  Everett                                     Osteopath   
...          ...      ...                                           ...   
54535     Angela  Everett                                     Osteopath   
54536     Kendra  Garrett                        Pension scheme manager   
54579  Alexandra    Young                                     Dietitian   
54589  Alexandra    Young                                     Dietitian   
54613     Kendra  Garrett                        Pension scheme manager   

                  phone                            email  client_id  order_id  \
5            5016915923         jessica.reyes@grimes.net      33615   1327730   
12           5016915923         jessica.reyes@grimes.net      33615   6059361   
137    444.233.9146x233            bmyers3696@wilson.com      46820   7109502   
214    444.233.9146x233            bmyers3696@wilson.com      46820   8967498   
216    729-620-6192x875           aeverett@rodriguez.com      66037   1203693   
...                 ...                              ...        ...       ...   
54535  729-620-6192x875           aeverett@rodriguez.com      66037   7550600   
54536      808.257.2029  garrettkendra8479@jefferson.com      24741   1886506   
54579  265-666-3155x118           ayoung9158@stevens.com      38378   8661147   
54589  265-666-3155x118           ayoung9158@stevens.com      38378   4508675   
54613      808.257.2029  garrettkendra8479@jefferson.com      24741   9114777   

       order_date  order_week  order_year  ... unit_cost unit_weight  qty  \
5      2023-05-02          18        2023  ...    154.44        2.79   78   
12     2023-05-16          20        2023  ...    198.38       27.51  568   
137    2023-05-12          19        2023  ...    111.47        5.65  118   
214    2023-02-06           6        2023  ...      6.26        7.05  133   
216    2023-04-09          14        2023  ...    166.09        0.26  157   
...           ...         ...         ...  ...       ...         ...  ...   
54535  2023-02-24           8        2023  ...     10.66        7.52  108   
54536  2023-05-26          21        2023  ...     40.05        4.70  721   
54579  2023-03-09          10        2023  ...     20.29       13.07  118   
54589  2023-03-25          12        2023  ...     38.26        0.70  685   
54613  2023-02-13           7        2023  ...     59.32       13.13   91   

       line_number  line_subtotal  total_weight  shipping_price  line_price  \
5                0       13481.52        217.62         1523.34    16392.81   
12               2      130492.32      15625.68       109379.76   262060.25   
137              9       16063.34        666.70         4666.90    22647.79   
214              3        1061.34        937.65         6563.55     8330.19   
216              2       47393.59         40.82          408.20    52223.46   
...            ...            ...           ...             ...         ...   
54535            2        1810.08        812.16         5685.12     8188.51   
54536            3       46014.22       3388.70        23720.90    76185.62   
54579            3        3687.50       1542.26        10795.82    15823.03   
54589            0       35784.40        479.50         3356.50    42761.43   
54613            2        8705.97       1194.83         8363.81    18648.73   

       line_cost  line_profit  
5       13569.66      2823.15  
12     222059.60     40000.65  
137     17820.36      4827.43  
214      7396.13       934.06  
216     26484.33     25739.13  
...          ...          ...  
54535    6836.40      1352.11  
54536  

In [80]:
client_summary

qty  shipping_price   line_price    line_cost  line_profit
client_id                                                               
24741      239862      5126448.37  82268892.02  45688899.71  36579992.31
38378       73667      3429455.40  12906550.87   9634720.98   3271829.89
66037       43018      1395151.85  10259514.79   7004482.98   3255031.81
46820       75768      1601448.84   9743794.36   7007191.64   2736602.72
33615       64313      1828984.89   8377308.52   6175313.91   2201994.61

In [81]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

client_summary_format = client_summary.copy()
client_summary_format['shipping_price'] /= 1e6
client_summary_format['line_price'] /= 1e6
client_summary_format['line_cost'] /= 1e6
client_summary_format['line_profit'] /= 1e6

client_summary_format = client_summary_format.rename(
    columns={
        'qty': 'Units',
        'shipping_price': 'Shipping',
        'line_price': 'Total Revenue',
        'line_cost': 'Total Cost',
        'line_profit': 'Total Profit'
    }
)

client_summary_format['Shipping'] = client_summary_format['Shipping'].apply(lambda x: f"${x:.2f}M")
client_summary_format['Total Revenue'] = client_summary_format['Total Revenue'].apply(lambda x: f"${x:.2f}M")
client_summary_format['Total Cost'] = client_summary_format['Total Cost'].apply(lambda x: f"${x:.2f}M")
client_summary_format['Total Profit'] = client_summary_format['Total Profit'].apply(lambda x: f"${x:.2f}M")


client_summary_format

Units Shipping Total Revenue Total Cost Total Profit
client_id                                                       
24741      239862   $5.13M       $82.27M    $45.69M      $36.58M
38378       73667   $3.43M       $12.91M     $9.63M       $3.27M
66037       43018   $1.40M       $10.26M     $7.00M       $3.26M
46820       75768   $1.60M        $9.74M     $7.01M       $2.74M
33615       64313   $1.83M        $8.38M     $6.18M       $2.20M

In [82]:
# Sort the updated data by "Total Profit" form highest to lowest
client_summary_format = client_summary_format.sort_values(by='Total Profit', ascending=False)
client_summary_format

Units Shipping Total Revenue Total Cost Total Profit
client_id                                                       
24741      239862   $5.13M       $82.27M    $45.69M      $36.58M
38378       73667   $3.43M       $12.91M     $9.63M       $3.27M
66037       43018   $1.40M       $10.26M     $7.00M       $3.26M
46820       75768   $1.60M        $9.74M     $7.01M       $2.74M
33615       64313   $1.83M        $8.38M     $6.18M       $2.20M